### DISCUSSION TITLE
Path tracing in O3DE

### DISCUSSION THREAD

**Discussion by ohmaya** [o3de/o3de#16184] [Show and tell]
<img width="720" alt="01" src="https://github.com/o3de/o3de/assets/5120856/cb129755-39dc-4312-8b93-df5f38c34633">
<img width="720" alt="02" src="https://github.com/o3de/o3de/assets/5120856/c8cf7352-3532-4083-8824-5e38cd80050b">
<img width="720" alt="03" src="https://github.com/o3de/o3de/assets/5120856/e1bdb19d-d7d3-4f30-b31e-9c1b6a620980">
<img width="720" alt="04" src="https://github.com/o3de/o3de/assets/5120856/8bbd16e9-006f-4001-b078-aaa2a5d27696">
<img width="720" alt="05" src="https://github.com/o3de/o3de/assets/5120856/b8434525-720b-4bc7-beb6-65f7d53198af">
<img width="628" alt="06" src="https://github.com/o3de/o3de/assets/5120856/f6283d85-16d3-4bd3-b8c1-834973e6c8b8">

They could reach 60 FPS (interactive frame rate, noisy initially) when the window is the current focus.

**Comment by aldenpang** [comment]
realtime path tracing? that insane! thanks for sharing, how did you do that?

**Reply by ohmaya** [reply]
After changing view perspective, initially it will be noisy. Will need accumulation for some, maybe many, frames for it to be smoother, so I think it still can only be considered interactive, but may **quickly get a sense of overall GI/reflections/sun soft shadows effects**.

**Reply by wsloi** [reply]
Nicely done!  Yep, that's temporal accumulation..  I'm just the light transport algorithm behind the curtain...

**Comment by aldenpang** [comment]
could you let me know your hardware configurartion? I guess you run it on PC with vulkan API

**Reply by ohmaya** [reply]
I am on D3D12 & RTX 3070.

The 1st image is very smooth and takes some time to reach that smoothness.

**Comment by moudgils** [comment]
Are you able to provide technical details related to the path tracer? Did you write one or integrate an existing one, etc? If you did integrate an existing one what api did you use to get access to scene details, etc?

**Reply by ohmaya** [reply]
It is more of an integration with some enhancements.

Scene details are provided by O3DE's RayTracingSceneSrg and RayTracingMaterialSrg.

**Reply by moudgils** [reply]
Nice. Are you planning on merging your changes at some point?

**Comment by ohmaya** [comment]
![DeNoiseAI-standard(PS)](https://github.com/o3de/o3de/assets/5120856/5a9abe8f-78b9-4ece-a782-239beba6dca2)

**Reply by VickyAtAZ** [reply]
Very nice. I wonder how many frames of accumulation it took to achieve this output? 
It would be nice if you can have a video showing the rendering when interact with scene.

**Reply by ohmaya** [reply]
According to NVIDIA, the only way to achieve quality real-time PT is using ReSTIR (current we are not). I found even NVIDIA Path Tracing SDK's real-time interior scene is not of high quality, so, PT does take time.

**Reply by ohmaya** [reply]
<img width="616" alt="StillMovingCamera" src="https://github.com/o3de/o3de/assets/5120856/97000499-8297-409b-9ffe-6c5262664827">

This is how noisy it is when the camera is still moving, giving an overall feel of the lighting.

**Comment by ohmaya** [comment]
![PinholeImaging](https://github.com/o3de/o3de/assets/5120856/a15956ae-535a-425b-8970-c8a35bd4a5f8)

Pinhole (triangular) imaging of the sun

**Comment by moudgils** [comment]
@ohmaya There are folks from VFX industry trying to improve O3DE for their use cases. A path tracer would be of great benefit. Have you considered contributing this work back? Thanks.

**Reply by ohmaya** [reply]
Would you start a dedicate project to implement PT (probably based on ReSTIR) for O3DE?

**Reply by moudgils** [reply]
Unfortunately I don have the capacity (although it is super interesting) to work on this project but if this work in a decent place, setup as a separate gem it would be cool to have other folks try it out.

**Reply by ohmaya** [reply]
I mean, maybe an official project could be started by recruiting some guys. That does not necessarily mean a separate Gem. While the PT in this post was not very mature and competitive to the latest technology.
